go-controller/pkg/factory/factory.go

```diff
@@ -182,14 +182,13 @@ func getQueueNum(oType reflect.Type, obj interface{}) uint32 {
 	return h.Sum32() % uint32(numEventQueues)
 }
 
-// enqueueEvent adds an event to the queue. Caller must hold at least a read lock
-// on the informer.
+// enqueueEvent adds an event to the queue.
 func (i *informer) enqueueEvent(oldObj, obj interface{}, processFunc func(*event)) {
 	i.RLock()
-	defer i.RUnlock()
-	queueIdx := getQueueNum(i.oType, obj)
-	if i.events[queueIdx] != nil {
-		i.events[queueIdx] <- &event{
+	queue := i.events[getQueueNum(i.oType, obj)]
```

Comment:
what happens if `removeHandler` gets called for this handler after the `Lock` has been relinquished? I fail to understand the dead-lock. Yes, the other readers will be blocked till the writer is able to relinquish the lock since it is blocked. But why are we not relinquishing the lock here after `enqueueEvent` is completed?

Reply:
I see the problem. The `forEachQueuedHandler` also seems to grab a RLock on the informer. The same lock that the enqueueEvent is holding.

Reply:
The deadlock is thus:
```
Thread 1:
i.RLock()
queue <- event

Thread 2:
i.RLock()
event <- queue

Thread 3:
i.Lock()
```

Thread 1 gets the lock, but the queue is full, because thread 2 can't get the lock. Thread 2 can't get the lock, because thread 3 has blocked all new `RLock()` from succeeding - this is a specific detail for how go RWMutexes work.

enqueueEevent doesn't act on handlers; it is called before handlers. So that's not a problem.

Arguably we don't need the lock at all! Given that we never ever touch the number of queues after the informer is created...

Reply:
So can we just get rid of the locking around `forEachQueuedHandler` and `enqueueEvent`?

Reply:
As I was looking at your change, I was wondering why you need to lock at all now and then saw your last comment. So if we don't need to use RLock at all can you please remove it?

Reply:
@squeed um, RLock() blocks Lock() though, no? In you example, if Thread 1 holds the lock then Thread 3 is blocked from getting the lock. So Thread 2 can proceed since it can run concurrently with Thread 1.

Reply:
> As I was looking at your change, I was wondering why you need to lock at all now and then saw your last comment. So if we don't need to use RLock at all can you please remove it?

@trozet we certainly need to lock becuase we need to handle shutting down the informer and removing handlers. And that requires setting the queues to nil, which requires locking on the queue array.

Reply:
> So can we just get rid of the locking around `forEachQueuedHandler` and `enqueueEvent`?

no, because a handler can be removed at any time and you cannot concurrently access the map when that's happening.

Reply:
We also need to lock so that added handlers get all updates. I'm not confident this is entirely correct right now anyways.

All of this code needs to go away or be really carefully audited.

Reply:
@dcbw RLock blocks Lock. But RWMutex also can decide for Lock() to block all subsequent RLocks() so that the Lock() has a chance to make progress.

Reply:
(It seems like people are confusing `i.handlers` and `i.events` in this thread.)